### PR TITLE
FI-346 Generator updates

### PR DESCRIPTION
**Pull request by arscan**
This makes a number of updates to the code generator:
* Changes to the name/arguments of the rake task to be more generic: `bundle exec rake:generate[*generator_name*,*implementation_guide_path*].
* Enhanced so that multiple versions of US core could exist simultaniously.  This is important because the US Core Guide now has both v3.0.0 and v.3.0.1, both categorized under "R4", and we want to be able to test each independently.  To run against v3.0.1, run `bundle exec rake inferno:generate[uscore,uscore_v3.0.1]`.
* Put full implementation guide materials in the repository (see uscore_v3.0.0 and uscore_v.3.0.1 directories).  The correct pattern for getting structure definitions is through implementation guide packages, and not through URLs, so I removed that.  This also means you'll need to download the package (generally a zip file) and update local files manually when updates come through.
* Updated to automatically run through a safe, layout-only rubocop autoformatter so we can make the code we write be indented properly for us to see, while the generator will auto-format the code that comes out.
* Changed the structure of the generator to be class-based, instead of being run as a external shell script.
* I added the version of the IG we are using embedded into the class name so that both v3.0.0 and v3.0.1 sequences could coexist.

This is mostly likely NOT the final form of the generator, as I think the `uscore` generator really should be more of a `patient_walk` or `read_only_resource_walk` or something so we could reuse it for other IGs.  But that would have been too much at once.

Test coverage is weak because of the above paragraph.  I think we are still figuring out how exactly we want to structure the generator, and hardening it with tests may be a bit of wasted effort at this point.

Pull requests into Inferno require the following items to be completed. Submitter and reviewer 
should check the relevant check boxes when the associated item is done. For items that are not 
applicable, note it's not applicable ("N/A") and check the box. For example, external Pull 
Requests do not require ticket links.

**Submitter:**
- [X] This pull request describes why these changes were made
- [X] Internal ticket for this PR: FI-346
- [X] Internal ticket links to this PR
- [X] Internal ticket is properly labeled (Community/Program)
- [X] Internal ticket has a justification for its Community/Program label
- [X] Code diff has been reviewed for extraneous/missing code
- [X] Tests are included and test edge cases: Note, generator immature so tests no added.
- [ ] Tests/code quality metrics have been run locally and pass

**Reviewer 1:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure
      where appropriate, and accomplishes the task's purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

**Reviewer 2:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure
      where appropriate, and accomplishes the task's purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

### CODE REVIEW

File: lib/app/modules/us_core_guidance/README.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+This folder is for guidance-based tests for US Core on FHIR R4.  This allows us
+to separate the autogenerated-tests from the manually created tests that are based
+from guidance text.  This solution could be improved.
```
Comment (Jammjammjamm) [inline lib/app/modules/us_core_guidance/README.md]:
```suggestion
on guidance text.  This solution could be improved.
```

File: lib/app/modules/uscore_v3.0.0/pediatric_bmi_for_age_sequence.rb [hunk 2]
```diff
@@ -44,8 +42,7 @@ def validate_resource_item(resource, property, value)
 
       details %(
 
-        The #{title} Sequence tests `#{title.gsub(/\s+/, '')}` resources associated with the provided patient.  The resources
-        returned will be checked for consistency against the [PediatricBmiForAge Argonaut Profile](http://hl7.org/fhir/us/core/StructureDefinition/pediatric-bmi-for-age)
+        The #{title} Sequence tests `#{title.gsub(/\s+/, '')}` resources associated with the provided patient.
```
Comment (Jammjammjamm) [inline lib/app/modules/uscore_v3.0.0/pediatric_bmi_for_age_sequence.rb]: Was this change intended?

**Comment by czh-orz** [comment]
As a note, I had to run bundle exec rake `bundle exec rake inferno:generate[uscore,uscore_v3.0.1]` to get it working.

File: generator/generator_base.rb [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+      def capability_statement(mode)
+        @resources_by_type['CapabilityStatement'].find { |re| re['rest'].any? { |r| r['mode'] == mode } }
+      end
+
+      def resource_path(resource)
+        "#{resource.class.name.demodulize}/#{resource.id}"
+      end
+
+      def format_output
+        # system('sh', "rubocop -x --display-only-fail-level-offenses #{sequence_out_path}/")
```
Comment (Jammjammjamm) [inline generator/generator_base.rb]: Is this comment needed?

File: generator/uscore/metadata_extractor.rb [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+module Inferno
+  module Generator
+    module USCoreMetadataExtractor
+      PROFILE_URIS = Inferno::ValidationUtil::US_CORE_R4_URIS
+
+      def profile_uri(profile)
+        "http://hl7.org/fhir/us/core/StructureDefinition/#{profile}"
+      end
+
+      def profile_json_uri(profile)
```
Comment (czh-orz) [inline generator/uscore/metadata_extractor.rb]: profile_json_uri isn't needed anymore

File: generator/uscore/metadata_extractor.rb [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+
+        # In case the profile doesn't start with US Core
+        class_name = "USCore#{reformatted_version}#{class_name}" unless class_name.start_with? 'USCore'
+
+        {
+          name: base_name.tr('-', '_'),
+          class_name: class_name,
+          resource: resource['type'],
+          profile: profile_uri(base_name), # link in capability statement is incorrect,
+          profile_json: profile_json_uri(base_name),
```
Comment (czh-orz) [inline generator/uscore/metadata_extractor.rb]: You can delete this entire line.

File: generator/uscore/metadata_extractor.rb [hunk 6]
```diff
@@ -0,0 +1,251 @@
...
+        "https://www.hl7.org/fhir/us/core/StructureDefinition-#{profile}.json"
+      end
+
+      def search_param_path(resource, param)
+        param = 'id' if param == '_id'
+        "SearchParameter/us-core-#{resource.downcase}-#{param}"
+      end
+
+      def extract_metadata
+        metadata = {
```
Comment (czh-orz) [inline generator/uscore/metadata_extractor.rb]: Do you know if they plan on updating the json for the capability statement and other definitions? The one offered from the site is still missing a lot of stuff.

**Comment by czh-orz** [comment]
I'm not sure if it's a good idea to publish the 3.0.1 version right now. The metadata is missing stuff like comparator expectation values and the us-core-include-provenance search parameter. That's why comparators are missing from sequences like pediatric bmi.

**Comment by arscan** [comment]
I created a separate ticket for the .gitattributes comment (FI-371) @Jammjammjamm

File: generator/generator_base.rb [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+        system("rubocop -x --display-only-fail-level-offenses #{sequence_out_path}")
+      end
+
+      def run
+        generate
+        format_output
+      end
+
+      # subclass must implement the following:
+      # def generate
```
[the remark this answers is not in the document]
Comment (arscan) [reply generator/generator_base.rb]: That's how I started, then I read [a blog opinion piece](http://chrisstump.online/2016/03/23/stop-abusing-notimplementederror/) how this was anti-pattern.  But honestly, I think that I'm more on your side now.

[the remark this answers is not in the document]
Comment (Jammjammjamm) [reply generator/generator_base.rb]: I agree about avoiding `NotImplementedError` in particular, which has different semantics. I think the expectation is that we won't be the only ones creating generators, so I think it's useful to be able to give a more useful error than method undefined.

File: generator/generator_base.rb [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+      end
+
+      def load_resources
+        @resource_by_path = Hash.new {}
+        @resources_by_type = Hash.new { |h, k| h[k] = [] }
+        Dir.glob("#{resource_file_path}/**/*.*") do |resource| # note one extra "*"
+          if File.file?(resource) && resource.end_with?('json', 'xml')
+
+            # We should consider using the native Ruby models instead of JSON
+            # There were problems with round-tripping certain SearchParameters though
```
[the remark this answers is not in the document]
Comment (arscan) [reply generator/generator_base.rb]: This should use the Ruby models instead of the json representation.  However, that would have been a pretty big lift on this PR, and I ran into some issues with SearchParamter extensions getting lost, and frankly the error message when you try to access a field that doesn't exist in a model produced tons of unuseful output.  I opened another ticket (FI-372)

File: lib/app/modules/uscore_v3.0.0/pediatric_bmi_for_age_sequence.rb (hunk 2 above)
Comment (arscan) [reply lib/app/modules/uscore_v3.0.0/pediatric_bmi_for_age_sequence.rb]: Yes.  This broke... something.  I'm not exactly sure what.  But this whole section should be a lot more specific.

File: generator/uscore/metadata_extractor.rb (hunk 6 above)
Comment (arscan) [reply generator/uscore/metadata_extractor.rb]: Grab all resources from the github repo directly.  For other things missing, we need to follow up on zulip.  You created tickets for this?

**Comment by arscan** [comment]
> I'm not sure if it's a good idea to publish the 3.0.1 version right now. The metadata is missing stuff like comparator expectation values and the us-core-include-provenance search parameter. That's why comparators are missing from sequences like pediatric bmi.

We need to get out in front of this to help make sure that the final published version is as correct as possible.  Perhaps this should go in infernotest but not show up on inferno.healthit.gov?

File: generator/generator_base.rb [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+            @resources_by_type[type].push(new_resource_json)
+          end
+        end
+      end
+
+      def ig_resource
+        @resources_by_type['ImplementationGuide'].first
+      end
+
+      def capability_statement(mode)
```
[the remark this answers is not in the document]
Comment (arscan) [reply generator/generator_base.rb]: I added a default here.

File: generator/generator_base.rb (hunk 7 above)
[the remark this answers is not in the document]
Comment (arscan) [reply generator/generator_base.rb]: I have this raise a `StandardError` now.

File: generator/generator_base.rb [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+        name = ig_metadata['name']
+        "#{version}#{name}"
+      end
+
+      def module_yml_out_path
+        File.expand_path('./lib/app/modules/')
+      end
+
+      def sequence_out_path
+        File.expand_path("./lib/app/modules/#{@path}")
```
[the remark this answers is not in the document]
Comment (arscan) [reply generator/generator_base.rb]: I updated this.

File: generator/generator_base.rb [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+        @resource_by_path = Hash.new {}
+        @resources_by_type = Hash.new { |h, k| h[k] = [] }
+        Dir.glob("#{resource_file_path}/**/*.*") do |resource| # note one extra "*"
+          if File.file?(resource) && resource.end_with?('json', 'xml')
+
+            # We should consider using the native Ruby models instead of JSON
+            # There were problems with round-tripping certain SearchParameters though
+            new_resource_json = JSON.parse(File.read(resource))
+            new_resource = FHIR.from_contents(File.read(resource))
+            @resource_by_path[resource_path(new_resource)] = new_resource_json
```
[the remark this answers is not in the document]
Comment (arscan) [reply generator/generator_base.rb]: I changed these to `attr_accessor`.  You may want to check to make sure that I did it in the way that you expect though.

File: generator/generator_base.rb [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+          end
+        end
+      end
+
+      def ig_resource
+        @resources_by_type['ImplementationGuide'].first
+      end
+
+      def capability_statement(mode)
+        @resources_by_type['CapabilityStatement'].find { |re| re['rest'].any? { |r| r['mode'] == mode } }
```
[the remark this answers is not in the document]
Comment (arscan) [reply generator/generator_base.rb]: I changed to `capability_statement_resource`, and moved to a multi-line block to give it a bit more breathing room.

File: lib/app/modules/us_core_guidance/README.md (hunk 1 above)
Comment (arscan) [reply lib/app/modules/us_core_guidance/README.md]: Fixed.

File: lib/app/modules/uscore_v3.0.0/pediatric_bmi_for_age_sequence.rb (hunk 2 above)
Comment (arscan) [reply lib/app/modules/uscore_v3.0.0/pediatric_bmi_for_age_sequence.rb]: I'm going to leave this as-is.  I think the whole details section needs to be a lot more descriptive anyhow (it should pull content out of the `description` field in the profiles, describe the test approach, etc).  I added a ticket for more robust documentation being generated here, see FI-375

File: generator/generator_base.rb (hunk 3 above)
Comment (arscan) [reply generator/generator_base.rb]: Nope. Removed.

File: generator/uscore/metadata_extractor.rb (hunk 4 above)
Comment (arscan) [reply generator/uscore/metadata_extractor.rb]: Good catch, removed.

File: generator/uscore/metadata_extractor.rb (hunk 5 above)
Comment (arscan) [reply generator/uscore/metadata_extractor.rb]: Yup, removed this as well, if we don't ever need this.

**Comment by arscan** [comment]
Great feedback @czh-orz, @Jammjammjamm. I believe I addressed everything here.

**Comment by czh-orz** [comment]
> > I'm not sure if it's a good idea to publish the 3.0.1 version right now. The metadata is missing stuff like comparator expectation values and the us-core-include-provenance search parameter. That's why comparators are missing from sequences like pediatric bmi.
> 
> We need to get out in front of this to help make sure that the final published version is as correct as possible. Perhaps this should go in infernotest but not show up on inferno.healthit.gov?

I'm okay with that. I'll make another ticket to update the 3.0.1 resources and regenerate so we don't forget about them.
